File: pipeline/kite9-pipeline-common/src/main/java/com/kite9/k9server/command/xml/move/Move.java
package com.kite9.k9server.command.xml.move;

import org.kite9.diagram.dom.elements.ADLDocument;
import org.w3c.dom.Element;

import com.kite9.k9server.adl.holder.pipeline.ADLDom;
import com.kite9.k9server.command.CommandException;
import com.kite9.k9server.command.xml.AbstractADLCommand;

public class Move extends AbstractADLCommand {

	public String moveId;  	// guy we are moving
	public String from, fromBefore, to, toBefore;
	
	@Override
	public Mismatch applyCommand(ADLDom in) throws CommandException {
		return move(in, moveId, from, fromBefore, to, toBefore);		
	}
	
	
	private Mismatch move(ADLDom in, String moveId, String from, String fromBefore, String to, String toBefore) {
		checkProperties();
		ADLDocument doc = in.getDocument();
		Element moveEl = doc.getElementById(moveId);
		Element toEl = doc.getElementById(to);
		Element toBeforeEl = doc.getElementById(toBefore);
		
		if ((moveEl == null) || (toEl == null)) {
			return () -> "Element missing, moveEl="+moveEl+", toEl="+toEl;
		}
		
		if (toBeforeEl == null) {
			toEl.appendChild(moveEl);
		} else {
			toEl.insertBefore(moveEl, toBeforeEl);
		}
		
		LOG.info("Completed move into "+to);
		return null;
	}

	
	protected void checkProperties() {
		ensureNotNull("moveId", moveId);
		ensureNotNull("from", from);
		ensureNotNull("to", to);
	}

	@Override
	public Mismatch undoCommand(ADLDom in) throws CommandException {
		return move(in, moveId, to, toBefore, from, fromBefore);		
	}



}

File: pipeline/kite9-pipeline-common/src/main/kotlin/com/kite9/k9server/command/xml/move/Move.kt
package com.kite9.k9server.command.xml.move

import com.kite9.k9server.adl.holder.pipeline.XMLDom.document
import org.kite9.diagram.logging.Kite9Log.send
import com.kite9.k9server.command.xml.AbstractADLCommand.ensureNotNull
import com.kite9.k9server.command.xml.AbstractADLCommand
import com.kite9.k9server.adl.holder.pipeline.ADLDom
import org.kite9.diagram.logging.Kite9Log
import com.kite9.k9server.command.Command.Mismatch

class Move : AbstractADLCommand() {
    // guy we are moving
    var moveId : String? = null
	var from: String? = null
    var fromBefore: String? = null
	var to: String? = null
	var toBefore: String? = null
    fun applyCommand(`in`: ADLDom, log: Kite9Log): Mismatch? {
        return move(`in`, moveId, to, toBefore, log)
    }

    private fun move(`in`: ADLDom, moveId: String?, to: String?, toBefore: String?, log: Kite9Log): Mismatch? {
        checkProperties()
        val doc = `in`.document
        val moveEl = doc.getElementById(moveId)
        val toEl = doc.getElementById(to)
        val toBeforeEl = doc.getElementById(toBefore)
        if (moveEl == null || toEl == null) {
            return Mismatch { "Element missing, moveEl=$moveEl, toEl=$toEl" }
        }
        if (toBeforeEl == null) {
            toEl.appendChild(moveEl)
        } else {
            toEl.insertBefore(moveEl, toBeforeEl)
        }
        log.send("Completed move into $to")
        return null
    }

    protected fun checkProperties() {
        ensureNotNull("moveId", moveId)
        ensureNotNull("from", from)
        ensureNotNull("to", to)
    }

    fun undoCommand(`in`: ADLDom, log: Kite9Log): Mismatch? {
        return move(`in`, moveId, from, fromBefore, log)
    }
}